src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs

```diff
@@ -1088,26 +1081,23 @@ void EmitBindCoreCall()
                     _ => $"{TypeIndex.GetParseMethodName(type)}({stringValueToParse_Expr}, {sectionPathExpr})",
                 };
 
-                if (!checkForNullSectionValue)
-                {
-                    InvokeWriteOnSuccess();
-                }
-                else
-                {
-                    EmitStartBlock($"if ({sectionValueExpr} is string {nonNull_StringValue_Identifier})");
-                    InvokeWriteOnSuccess();
-                    EmitEndBlock();
-                }
+                if (checkForNullSectionValue) EmitStartBlock($"if ({sectionValueExpr} is string {nonNull_StringValue_Identifier})");
+                writeOnSuccess?.Invoke(parsedValueExpr);
+                if (checkForNullSectionValue) EmitEndBlock();
 
-                if (useDefaultValueIfSectionValueIsNull)
+                if (defaultValueSource is not null)
                 {
-                    parsedValueExpr = $"default";
+                    Debug.Assert(checkForNullSectionValue);
+
                     EmitStartBlock($"else if (defaultValueIfNotFound)");
-                    InvokeWriteOnSuccess();
+                    _writer.WriteLine($"var currentValue = {defaultValueSource};");
+                    // `is object` will be happy with value types, but will give CS0183 which we suppress
+                    // `is null` gives "Cannot convert null to 'int' because it is a non-nullable value type"
+                    EmitStartBlock($"if (currentValue is object) ");
```

Comment:
Is it possible passing something telling if we have a value type and avoid the check and disabling the warning at that time?

Reply:
It should be possible, yes.  I spent a while on it, but the complexity it introduced led me to the current solution.  If better generated code for this case is desired, a general refactor of this might be worthwhile.

The `is object` check for value types [is compiled to IL as a constant load](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANmgExAGoAfAAQCYBGAWAChyBmAAipYGEWBvBl/ltgB2GQQG4GfAczYoWAWQAUAShZT+vegO2CAZi0XZBAZxYRgAKxhgMy9Ts06nbagE5FAIgxQArjA/KElrOAL72YfQhQA==) and the [branch is eliminated in JIT](https://godbolt.org/z/8TPvf3b8e), so I did not pursue the issue further.  It does look like it the check is still around in [AOT](https://godbolt.org/z/T1s7EfW49).

Reply:
I am not worried about the performance. I am trying to avoid global suppression of warnings which in the future can hide something we are not aware of.